rust/src/lib.rs

```diff
@@ -1248,41 +1304,68 @@ pub extern "C" fn zcashlc_vec_string_free(v: *mut *mut c_char, len: usize, capac
     };
 }
 
-/// TEST TEST 123 TEST
-///
-///
-
-/// Derives a transparent private key from seed
+/// Derives a transparent account private key from seed
 #[no_mangle]
-pub unsafe extern "C" fn zcashlc_derive_transparent_private_key_from_seed(
+pub unsafe extern "C" fn zcashlc_derive_transparent_account_private_key_from_seed(
     seed: *const u8,
     seed_len: usize,
     account: i32,
-    index: i32,
     network_id: u32,
 ) -> *mut c_char {
     let res = catch_panic(|| {
         let network = parse_network(network_id)?;
         let seed = slice::from_raw_parts(seed, seed_len);
         let account = if account >= 0 {
-            account as u32
+            AccountId::from(account as u32)
         } else {
             return Err(format_err!("account argument must be positive"));
         };
 
-        let index = if index >= 0 {
-            index as u32
-        } else {
-            return Err(format_err!("index argument must be positive"));
-        };
-        let sk = derive_secret_key_from_seed(&network, &seed, AccountId(account), index).unwrap();
-        let sk_wif = Wif::from_secret_key(&sk, true);
+        // Derive the USK to ensure it exists, and fetch its transparent component.
+        let usk = UnifiedSpendingKey::from_seed(&network, &seed, account)
+            .map_err(|e| format_err!("error generating unified spending key from seed: {:?}", e))?;
+        // Derive the corresponding BIP 32 extended privkey.
+        let xprv =
+            p2pkh_xprv(&network, &seed, account).expect("USK derivation should ensure this exists");
+        // Verify that we did derive the same privkey.
+        assert_eq!(
+            usk.transparent().to_account_pubkey().serialize(),
+            legacy::keys::AccountPrivKey::from_extended_privkey(xprv.extended_key.clone())
+                .to_account_pubkey()
+                .serialize(),
+        );
+        // Encode using the BIP 32 xprv serialization format.
+        let xprv_str: String = xprv.serialize();
 
-        Ok(CString::new(sk_wif.0.to_string()).unwrap().into_raw())
+        Ok(CString::new(xprv_str).unwrap().into_raw())
     });
     unwrap_exc_or_null(res)
 }
 
+fn p2pkh_xprv<P: Parameters>(
```

Comment:
Why is this not using https://github.com/zcash/librustzcash/blob/main/zcash_primitives/src/legacy/keys.rs#L24 ?

Reply:
I needed specifically the `hdwallet_bitcoin::PrivKey` type here, and while that does store an `hdwallet::ExtendedPrivKey` internally, and the helper you're referencing returns a newtype around `hdwallet::ExtendedPrivKey`, I don't have any way to access it (and I'm not sure I want one in the public API of the Rust crates just yet). What we really need is serialization APIs on `AccountPrivKey`, which could then replace this logic.

Reply:
The main thing that I don't care for is having the derivation path logic reimplemented, but for now I guess that's okay.